File: base/src/main/groovy/com/knight/plugin/doublecheck/java/ASMUtils.java
package com.knight.plugin.doublecheck.java;

import org.objectweb.asm.Type;

import static org.objectweb.asm.Opcodes.*;

public class ASMUtils {

    static boolean isPrivate(int access) {
        return (access & ACC_PRIVATE) != 0;
    }

    static boolean isPublic(int access) {
        return (access & ACC_PUBLIC) != 0;
    }

    static boolean isStatic(int access) {
        return (access & ACC_STATIC) != 0;
    }

    static String convertSignature(String name, String desc) {
        Type method = Type.getType(desc);
        StringBuilder sb = new StringBuilder();
        sb.append(method.getReturnType().getClassName()).append(" ").append(name);
        sb.append("(");
        for (int i = 0; i < method.getArgumentTypes().length; i++) {
            sb.append(method.getArgumentTypes()[i].getClassName());
            if (i != method.getArgumentTypes().length - 1) {
                sb.append(",");
            }
        }
        sb.append(")");
        return sb.toString();
    }
}

File: knight-transform/src/main/kotlin/com/knight/transform/Utils/ASMUtils.kt
package transform.Utils

import org.objectweb.asm.Type

import org.objectweb.asm.Opcodes.ACC_PRIVATE
import org.objectweb.asm.Opcodes.ACC_PUBLIC
import org.objectweb.asm.Opcodes.ACC_STATIC

object ASMUtils {
    fun isPrivate(access: Int): Boolean {
        return access and ACC_PRIVATE != 0
    }

    fun isPublic(access: Int): Boolean {
        return access and ACC_PUBLIC != 0
    }

    fun isStatic(access: Int): Boolean {
        return access and ACC_STATIC != 0
    }

    fun convertSignature(name: String, desc: String): String {
        val method = Type.getType(desc)
        val sb = StringBuilder()
        sb.append(method.returnType.className).append(" ").append(name)
        sb.append("(")
        for (i in 0 until method.argumentTypes.size) {
            sb.append(method.argumentTypes[i].className)
            if (i != method.argumentTypes.size - 1) {
                sb.append(",")
            }
        }
        sb.append(")")
        return sb.toString()
    }
}